File: list/src/main/java/LinkedList.java
public class LinkedList implements List {
    private Node head;
    private int size;
    LinkedList(){
        head = new Node(null,null);
        size = 0;
    }
    public void clear() {
        for(Node p = head;p.next!=null;p=p.next){
            Node q = p.next;
            q.next = null;
            q.data = null;
        }
        size = 0;
    }

    public boolean isEmpty() {
        if(size == 0){
            return true;
        }
        return false;
    }

    public int length() {
        return size;
    }

    public Object get(int index) {
        if(index <0 || index >=size){
            throw new IndexOutOfBoundsException("illegal index:"+index+">"+size);
        }
        Node p = head;
        for(int i=0;i<=index;i++){
            p = p.next;
        }
        return p.data;
    }

    public int indexOf(Object o) {
        Node p = head.next;
        for(int i = 0;i < size;i++){
            if(p.data.equals(o)){
                return i;
            }
            p = p.next;
        }
        return -1;
    }

    public Object pre(Object o) {
        int index = indexOf(o);
        if(index == -1 ||index == 0){
            return null;
        }else {
            return get(index - 1);
        }
    }

    public Object next(Object o) {
        int index = indexOf(o);
        if(index == -1 || index == 0 || index >= size-1){
            return null;
        }else{
            return get(index+1);
        }
    }

    public void insert(int index, Object o) {
        Node p = head;
        if(index < 0 || index > size){
            throw new IndexOutOfBoundsException("illegal index:"+index+"{"+size+"}");
        }
        for(int i = 0; i<=index -1;i++){
            p = p.next;
        }

        Node q = new Node(o, null);
        q.next = p.next;
        p.next = q;

        size++;
    }

    public void delete(int index) {
        Node p =head;
        if(index < 0 || index >= size){
            throw new IndexOutOfBoundsException("illegal index:"+index+"{"+size+"}");
        }
        for(int i = 0; i<=index -1;i++){
            p = p.next;
        }
        Node back = p.next;
        p.next = p.next.next;
        size--;
    }

    public void display() {
        int i = 0;
        for(Node p = head.next;i<size;p=p.next,i++){
            System.out.print(p.data+" ");
        }
        System.out.println();
    }
    public Node getHead(){
        return head;
    }

    public void union(List lb) {
        Node pa = head;
        Node pb = ((LinkedList)lb).getHead().next;
        Node pc;
        while(pb!=null){
            if(this.indexOf(pb.data)==-1){
                pc = pb.next;
                pb.next = pa.next;
                pa.next = pb;
                pb = pc;
                pa = pa.next;
                size++;
            }else{
                pb = pb.next;
            }
        }
    }

    public List merge(List lb) {
        Node pa = head.next;
        Node pb = ((LinkedList)lb).getHead().next;
        Node pc = head;
        while(pa !=null && pb !=null){
            if((Integer)pa.data <= (Integer)pb.data){
                pc.next = pa;
                pc = pa;
                pa = pa.next;
                size++;
            }else{
                pc.next = pb;
                pc = pb;
                pb = pb.next;
            }
        }

        pc.next = pa!=null?pa:pb;
        return null;
    }
}

class Node{
     Object data;
     Node next;
     Node(Object data, Node next){
         this.data = data;
         this.next = next;
     }

}

File: Python/list/LinkedList.py
# -*- coding: UTF-8 -*-


class Node:
    def __init__(self, prev, item, next):
        self.prev = prev
        self.item = item
        self.next = next


class LinkedList:
    def __init__(self):
        self.__head = Node(None, None, None)
        self.__tail = Node(self.__head,None,None)
        self.__head.next = self.__tail
        self.__first = self.__last = None
        self.__size = 0

    def __node(self, item):
        if self.__size == 0:
            return None
        p = self.__first
        while p != self.__tail:
            if p.item == item:
                return p
            p = p.next
        return None

    def __unlink(self, node):
        node.prev.next = node.next
        node.next.pre = node.prev
        if node == self.__last:
            self.__last = self.__last.prev
        if node == self.__first:
            self.__first = self.__first.next
        if node == self.__first == self.__last:
            self.__first = self.__last = None

    def add(self, item):
        p = Node(self.__last, item, self.__tail)
        if p.prev is None:
            self.__first = p
            self.__last = p
            p.prev = self.__head
            self.__head.next = p
            self.__tail.prev = p
        else:
            self.__tail.prev = p
            self.__last.next = p
            self.__last = p
        self.__size += 1

    def index_of(self, item):
        p = self.__first
        if p is None:
            return -1
        index = 0
        while p != self.__tail:
            if p.item == item:
                return index
            p = p.next
            index += 1
        return -1

    def remove(self, item):
        if self.__size == 0:
            return
        self.__unlink(self.__node(item))
        self.__size -= 1

    def insert(self, index, item):
        if index < 0 or index > self.__size:
            return
        p = self.__first
        for i in range(index):
            p = p.next
        n = Node(p.prev, item, p)
        if p == self.__first:
            self.__head.next = n
            p.prev = n
            self.__first = n
        elif p == self.__tail:
            self.__tail.prev.next = n
            self.__tail.prev = n
            self.__last = n
        else:
            p.prev.next = n
            p.prev = n

        self.__size += 1

    def get(self, index):
        if index < 0 or index >= self.__size:
            return None
        p = self.__first
        for i in range(index):
            p = p.next
        return p.item

    def size(self):
        return self.__size

    def get_first(self):
        return self.__first.item

    def get_last(self):
        return self.__last.item

    def prev(self, item):
        index = self.index_of(item)
        if index == -1 or index == 0:
            return None
        return self.__node(item).prev.item

    def succ(self, item):
        index = self.index_of(item)
        if index == -1 or index == self.__size - 1:
            return None
        return self.__node(item).next.item

    def display(self):
        p = self.__first
        while p.next != self.__tail:
            print(p.item, end="->")

            p = p.next
        print(p.item)

    def clear(self):
        ne = p = self.__first
        while ne != self.__tail:
            ne = p.next
            p.item = None
            p.prev = None
            p.next = None
        self.__size = 0
        self.__first = self.__last = None
        self.__head.next = self.__tail
        self.__tail.prev = self.__head


if __name__ == '__main__':
    ll = LinkedList()
    ll.add(1)
    print("链表的大小:", end="")
    print(ll.size(), end="\n")
    print("链表的头:", end="")
    print(ll.get_first(), end="\n")
    print("链表的尾", end="")
    print(ll.get_last(), end="\n")
    print("1的下标:", end='')
    print(ll.index_of(1), end='\n')
    ll.display()
    print()

    ll.insert(0, 2)
    print("链表的大小:", end="")
    print(ll.size(), end="\n")
    print("链表的头:", end="")
    print(ll.get_first(), end="\n")
    print("链表的尾:", end="")
    print(ll.get_last(), end="\n")
    print("2的下标:", end='')
    print(ll.index_of(2), end='\n')
    ll.display()
    print()

    ll.insert(2, 3)
    print("链表的大小:", end="")
    print(ll.size(), end="\n")
    print("链表的头:", end="")
    print(ll.get_first(), end="\n")
    print("链表的尾", end="")
    print(ll.get_last(), end="\n")
    print("3的下标:", end='')
    print(ll.index_of(3), end='\n')
    ll.display()
    print()

    ll.insert(1, 4)
    print("链表的大小:", end="")
    print(ll.size(), end="\n")
    print("链表的头:", end="")
    print(ll.get_first(), end="\n")
    print("链表的尾", end="")
    print(ll.get_last(), end="\n")
    print("3的下标:", end='')
    print(ll.index_of(3), end='\n')
    print("2的前驱:", end='')
    print(ll.prev(2), end='\n')
    print("3的后继:", end='')
    print(ll.succ(3), end='\n')
    print("4的前驱:", end='')
    print(ll.prev(4), end='\n')
    print("4的后继:", end='')
    print(ll.succ(4), end='\n')
    ll.display()
    print()
